File: services/storm-exploration/app/domain/contracts.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, HttpUrl, model_validator


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)


class TimeRange(StrictModel):
    from_year: int | None = Field(default=None, alias="fromYear", ge=1800, le=2200)
    to_year: int | None = Field(default=None, alias="toYear", ge=1800, le=2200)

    @model_validator(mode="after")
    def validate_order(self) -> "TimeRange":
        if self.from_year is not None and self.to_year is not None and self.from_year > self.to_year:
            raise ValueError("fromYear cannot be after toYear")
        return self


class ExplorationScope(StrictModel):
    time_range: TimeRange | None = Field(default=None, alias="timeRange")
    disciplines: list[str] = Field(default_factory=list, max_length=20)
    excluded_topics: list[str] = Field(default_factory=list, alias="excludedTopics", max_length=50)


class SourcePolicy(StrictModel):
    use_web: bool = Field(alias="useWeb")
    use_user_documents: bool = Field(alias="useUserDocuments")
    user_resource_ids: list[str] = Field(alias="userResourceIds", max_length=200)

    @model_validator(mode="after")
    def validate_sources(self) -> "SourcePolicy":
        if not self.use_web and not self.use_user_documents:
            raise ValueError("at least one source must be enabled")
        if not self.use_user_documents and self.user_resource_ids:
            raise ValueError("userResourceIds require useUserDocuments=true")
        return self


class ExplorationLimits(StrictModel):
    max_perspectives: int = Field(alias="maxPerspectives", ge=1, le=12)
    max_questions_per_perspective: int = Field(alias="maxQuestionsPerPerspective", ge=1, le=12)
    max_search_queries: int = Field(alias="maxSearchQueries", ge=1, le=200)
    max_sources: int = Field(alias="maxSources", ge=1, le=500)
    maximum_wall_time_ms: int = Field(alias="maximumWallTimeMs", ge=10_000, le=3_600_000)
    maximum_model_calls: int = Field(alias="maximumModelCalls", ge=1, le=500)
    maximum_inspection_count: int = Field(alias="maximumInspectionCount", ge=1, le=100)


class ModelProfile(StrictModel):
    provider: str = Field(min_length=1, max_length=160)
    model: str = Field(min_length=1, max_length=160)
    reasoning_effort: Literal["none", "low", "medium"] = Field(default="none", alias="reasoningEffort")


class ExplorationRequest(StrictModel):
    schema_version: Literal["storm-exploration-request-v1"] = Field(alias="schemaVersion")
    exploration_id: str = Field(alias="explorationId", min_length=1, max_length=160)
    topic: str = Field(min_length=3, max_length=1_000)
    purpose: Literal[
        "literature_review",
        "grant_topic_exploration",
        "field_landscape",
        "technology_comparison",
    ]
    language: Literal["zh", "en"]
    scope: ExplorationScope = Field(default_factory=ExplorationScope)
    source_policy: SourcePolicy = Field(alias="sourcePolicy")
    limits: ExplorationLimits
    model_profile: ModelProfile = Field(alias="modelProfile")
    user_resource_snapshot_hash: str | None = Field(
        default=None,
        alias="userResourceSnapshotHash",
        pattern=r"^[a-fA-F0-9]{64}$",
    )

    @model_validator(mode="after")
    def validate_user_snapshot(self) -> "ExplorationRequest":
        if self.source_policy.use_user_documents and not self.user_resource_snapshot_hash:
            raise ValueError("user-document exploration requires a frozen snapshot hash")
        return self


ExplorationStatus = Literal[
    "queued",
    "running",
    "partial",
    "complete",
    "failed",
    "unknown_outcome",
    "expired",
    "cancelled",
]


class FailurePayload(StrictModel):
    code: str = Field(min_length=1, max_length=160)
    category: Literal["contract", "provider", "infrastructure", "unknown_outcome"]
    retryability: Literal["none", "safe", "unknown"]
    technical_message: str = Field(alias="technicalMessage", min_length=1, max_length=8_000)
    user_message_code: str = Field(alias="userMessageCode", min_length=1, max_length=160)


class StartResponse(StrictModel):
    remote_execution_id: str = Field(alias="remoteExecutionId", min_length=1, max_length=160)
    status: Literal["queued", "running"]
    next_check_at: datetime | None = Field(default=None, alias="nextCheckAt")


class StatusResponse(StrictModel):
    remote_execution_id: str = Field(alias="remoteExecutionId", min_length=1, max_length=160)
    status: ExplorationStatus
    phase: str = Field(min_length=1, max_length=160)
    result_location: str | None = Field(default=None, alias="resultLocation", max_length=2_000)
    next_check_at: datetime | None = Field(default=None, alias="nextCheckAt")
    failure: FailurePayload | None = None


class PerspectiveCandidate(StrictModel):
    key: str = Field(min_length=1, max_length=160)
    title: str = Field(min_length=1, max_length=4_000)
    rationale: str = Field(min_length=1, max_length=12_000)


class QuestionCandidate(StrictModel):
    key: str = Field(min_length=1, max_length=160)
    perspective_key: str = Field(alias="perspectiveKey", min_length=1, max_length=160)
    question: str = Field(min_length=1, max_length=4_000)
    importance: Literal["high", "medium", "low"] = "medium"
    follow_ups: list[str] = Field(default_factory=list, alias="followUps", max_length=20)


class SearchCandidate(StrictModel):
    key: str = Field(min_length=1, max_length=160)
    question_key: str = Field(alias="questionKey", min_length=1, max_length=160)
    query: str = Field(min_length=1, max_length=4_000)
    source_type: Literal["web", "user_corpus"] = Field(alias="sourceType")


class SourceCandidate(StrictModel):
    key: str = Field(min_length=1, max_length=160)
    title: str = Field(min_length=1, max_length=4_000)
    url: HttpUrl | None = None
    doi: str | None = Field(default=None, max_length=300)
    authors: list[str] = Field(default_factory=list, max_length=100)
    year: int | None = Field(default=None, ge=1000, le=2200)
    snippet: str | None = Field(default=None, max_length=12_000)
    retrieved_by: str = Field(alias="retrievedBy", min_length=1, max_length=160)


class KnowledgeCandidate(StrictModel):
    key: str = Field(min_length=1, max_length=160)
    title: str = Field(min_length=1, max_length=4_000)
    summary: str = Field(min_length=1, max_length=12_000)
    parent_key: str | None = Field(default=None, alias="parentKey", max_length=160)
    source_keys: list[str] = Field(default_factory=list, alias="sourceKeys", max_length=100)


class OutlineSectionCandidate(StrictModel):
    heading: str = Field(min_length=1, max_length=4_000)
    purpose: str = Field(min_length=1, max_length=12_000)
    question_keys: list[str] = Field(default_factory=list, alias="questionKeys", max_length=100)
    source_keys: list[str] = Field(default_factory=list, alias="sourceKeys", max_length=100)


class OutlineCandidate(StrictModel):
    key: str = Field(min_length=1, max_length=160)
    title: str = Field(min_length=1, max_length=4_000)
    sections: list[OutlineSectionCandidate] = Field(min_length=1, max_length=40)


class ProviderCallEvidence(StrictModel):
    provider: str = Field(min_length=1, max_length=160)
    kind: Literal["model", "search"]
    operation: str = Field(min_length=1, max_length=160)
    model: str | None = Field(default=None, max_length=160)
    provider_request_id: str | None = Field(
        default=None, alias="providerRequestId", max_length=500
    )
    status: Literal["succeeded", "failed", "unknown_outcome"]
    input_tokens: int | None = Field(default=None, alias="inputTokens", ge=0)
    output_tokens: int | None = Field(default=None, alias="outputTokens", ge=0)
    estimated_cost_usd: float | None = Field(
        default=None, alias="estimatedCostUsd", ge=0
    )
    started_at: datetime = Field(alias="startedAt")
    finished_at: datetime = Field(alias="finishedAt")


class UsagePayload(StrictModel):
    model_calls: int = Field(default=0, alias="modelCalls", ge=0)
    search_calls: int = Field(default=0, alias="searchCalls", ge=0)
    input_tokens: int | None = Field(default=None, alias="inputTokens", ge=0)
    output_tokens: int | None = Field(default=None, alias="outputTokens", ge=0)
    estimated_cost_usd: float | None = Field(default=None, alias="estimatedCostUsd", ge=0)
    provider_calls: list[ProviderCallEvidence] = Field(
        default_factory=list, alias="providerCalls", max_length=500
    )


class ExplorationResult(StrictModel):
    schema_version: Literal["storm-exploration-result-v1"] = Field(alias="schemaVersion")
    exploration_id: str = Field(alias="explorationId", min_length=1, max_length=160)
    status: Literal["complete", "partial", "failed"]
    perspectives: list[PerspectiveCandidate] = Field(default_factory=list, max_length=20)
    questions: list[QuestionCandidate] = Field(default_factory=list, max_length=200)
    searches: list[SearchCandidate] = Field(default_factory=list, max_length=500)
    sources: list[SourceCandidate] = Field(default_factory=list, max_length=1_000)
    knowledge: list[KnowledgeCandidate] = Field(default_factory=list, max_length=1_000)
    outlines: list[OutlineCandidate] = Field(default_factory=list, max_length=20)
    unresolved_questions: list[str] = Field(default_factory=list, alias="unresolvedQuestions", max_length=200)
    warnings: list[str] = Field(default_factory=list, max_length=200)
    usage: UsagePayload = Field(default_factory=UsagePayload)

File: lib/research-exploration/contracts.ts
import { z } from "zod";

const IdentifierSchema = z.string().trim().min(1).max(160);
const TextSchema = z.string().trim().min(1).max(4_000);
const DateTimeSchema = z.iso.datetime({ offset: true });
const Sha256Schema = z.string().regex(/^[a-f0-9]{64}$/i);

export const ResearchExplorationPurposeSchema = z.enum([
  "literature_review",
  "grant_topic_exploration",
  "field_landscape",
  "technology_comparison",
]);

export const ResearchExplorationModeSchema = z.enum([
  "off",
  "advisory",
  "required",
]);

export const ResearchExplorationVersionSnapshotSchema = z
  .object({
    packageVersion: IdentifierSchema,
    adapterVersion: IdentifierSchema,
    outputContractVersion: IdentifierSchema,
    promptConfigurationVersion: IdentifierSchema,
  })
  .strict();

export const ResearchExplorationInputSchema = z
  .object({
    explorationId: IdentifierSchema,
    topic: z.string().trim().min(3).max(1_000),
    purpose: ResearchExplorationPurposeSchema,
    language: z.enum(["zh", "en"]),
    scope: z
      .object({
        timeRange: z
          .object({
            fromYear: z.number().int().min(1800).max(2200).optional(),
            toYear: z.number().int().min(1800).max(2200).optional(),
          })
          .strict()
          .superRefine((range, context) => {
            if (
              range.fromYear !== undefined &&
              range.toYear !== undefined &&
              range.fromYear > range.toYear
            ) {
              context.addIssue({
                code: "custom",
                path: ["fromYear"],
                message: "Research start year cannot be after the end year.",
              });
            }
          })
          .optional(),
        disciplines: z.array(TextSchema).max(20).default([]),
        excludedTopics: z.array(TextSchema).max(50).default([]),
      })
      .strict()
      .default({ disciplines: [], excludedTopics: [] }),
    sourcePolicy: z
      .object({
        useWeb: z.boolean(),
        useUserDocuments: z.boolean(),
        userResourceIds: z.array(IdentifierSchema).max(200),
      })
      .strict()
      .superRefine((policy, context) => {
        if (!policy.useWeb && !policy.useUserDocuments) {
          context.addIssue({
            code: "custom",
            message: "At least one research source must be enabled.",
          });
        }
        if (!policy.useUserDocuments && policy.userResourceIds.length > 0) {
          context.addIssue({
            code: "custom",
            path: ["userResourceIds"],
            message: "User resource IDs require useUserDocuments=true.",
          });
        }
      }),
    limits: z
      .object({
        maxPerspectives: z.number().int().min(1).max(12),
        maxQuestionsPerPerspective: z.number().int().min(1).max(12),
        maxSearchQueries: z.number().int().min(1).max(200),
        maxSources: z.number().int().min(1).max(500),
        maximumWallTimeMs: z.number().int().min(10_000).max(3_600_000),
        maximumModelCalls: z.number().int().min(1).max(500),
        maximumInspectionCount: z.number().int().min(1).max(100),
      })
      .strict(),
    modelProfile: z
      .object({
        provider: IdentifierSchema,
        model: IdentifierSchema,
        reasoningEffort: z.enum(["none", "low", "medium"]).default("none"),
      })
      .strict(),
    userResourceSnapshotHash: Sha256Schema.optional(),
  })
  .strict()
  .superRefine((input, context) => {
    if (input.sourcePolicy.useUserDocuments && !input.userResourceSnapshotHash) {
      context.addIssue({
        code: "custom",
        path: ["userResourceSnapshotHash"],
        message: "User-document exploration requires a frozen resource snapshot hash.",
      });
    }
  });

const PerspectiveCandidateSchema = z
  .object({
    perspectiveKey: IdentifierSchema,
    title: TextSchema,
    rationale: z.string().trim().min(1).max(8_000),
  })
  .strict();

const ResearchQuestionCandidateSchema = z
  .object({
    questionKey: IdentifierSchema,
    perspectiveKey: IdentifierSchema,
    question: TextSchema,
    importance: z.enum(["high", "medium", "low"]),
    followUpQuestions: z.array(TextSchema).max(20),
  })
  .strict();

const SearchPlanCandidateSchema = z
  .object({
    queryKey: IdentifierSchema,
    questionKey: IdentifierSchema,
    query: TextSchema,
    sourceType: z.enum(["web", "user_corpus"]),
  })
  .strict();

const SourceCandidateSchema = z
  .object({
    sourceCandidateKey: IdentifierSchema,
    title: TextSchema,
    url: z.url().optional(),
    doi: z.string().trim().min(3).max(300).optional(),
    authors: z.array(z.string().trim().min(1).max(300)).max(100).optional(),
    year: z.number().int().min(1000).max(2200).optional(),
    snippet: z.string().trim().min(1).max(12_000).optional(),
    retrievedBy: IdentifierSchema,
  })
  .strict();

const KnowledgeNodeCandidateSchema = z
  .object({
    nodeKey: IdentifierSchema,
    title: TextSchema,
    summary: z.string().trim().min(1).max(12_000),
    parentNodeKey: IdentifierSchema.optional(),
    supportingSourceCandidateKeys: z.array(IdentifierSchema).max(100),
  })
  .strict();

const OutlineSectionCandidateSchema = z
  .object({
    heading: TextSchema,
    purpose: z.string().trim().min(1).max(8_000),
    questionKeys: z.array(IdentifierSchema).max(100),
    supportingSourceCandidateKeys: z.array(IdentifierSchema).max(100),
  })
  .strict();

const OutlineCandidateSchema = z
  .object({
    outlineKey: IdentifierSchema,
    title: TextSchema,
    sections: z.array(OutlineSectionCandidateSchema).min(1).max(40),
  })
  .strict();

function duplicateValues(values: ReadonlyArray<string>): string[] {
  const seen = new Set<string>();
  const duplicates = new Set<string>();
  for (const value of values) {
    if (seen.has(value)) duplicates.add(value);
    seen.add(value);
  }
  return [...duplicates];
}

export const ResearchExplorationProposalSchema = z
  .object({
    schemaVersion: z.literal(1),
    explorationId: IdentifierSchema,
    status: z.enum(["complete", "partial", "failed"]),
    perspectives: z.array(PerspectiveCandidateSchema).max(20),
    researchQuestions: z.array(ResearchQuestionCandidateSchema).max(200),
    searchPlans: z.array(SearchPlanCandidateSchema).max(500),
    sourceCandidates: z.array(SourceCandidateSchema).max(1_000),
    knowledgeNodes: z.array(KnowledgeNodeCandidateSchema).max(1_000),
    outlineCandidates: z.array(OutlineCandidateSchema).max(20),
    unresolvedQuestions: z.array(TextSchema).max(200),
    warnings: z.array(z.string().trim().min(1).max(4_000)).max(200),
    usage: z
      .object({
        modelCalls: z.number().int().nonnegative(),
        searchCalls: z.number().int().nonnegative(),
        inputTokens: z.number().int().nonnegative().optional(),
        outputTokens: z.number().int().nonnegative().optional(),
        estimatedCostUsd: z.number().nonnegative().optional(),
      })
      .strict(),
  })
  .strict()
  .superRefine((proposal, context) => {
    const keyedCollections = [
      ["perspectives", proposal.perspectives.map((item) => item.perspectiveKey)],
      ["researchQuestions", proposal.researchQuestions.map((item) => item.questionKey)],
      ["searchPlans", proposal.searchPlans.map((item) => item.queryKey)],
      ["sourceCandidates", proposal.sourceCandidates.map((item) => item.sourceCandidateKey)],
      ["knowledgeNodes", proposal.knowledgeNodes.map((item) => item.nodeKey)],
      ["outlineCandidates", proposal.outlineCandidates.map((item) => item.outlineKey)],
    ] as const;
    for (const [path, values] of keyedCollections) {
      for (const duplicate of duplicateValues(values)) {
        context.addIssue({
          code: "custom",
          path: [path],
          message: `Duplicate candidate key: ${duplicate}`,
        });
      }
    }
    const perspectiveKeys = new Set(
      proposal.perspectives.map((item) => item.perspectiveKey),
    );
    const questionKeys = new Set(
      proposal.researchQuestions.map((item) => item.questionKey),
    );
    const sourceKeys = new Set(
      proposal.sourceCandidates.map((item) => item.sourceCandidateKey),
    );
    const knowledgeKeys = new Set(
      proposal.knowledgeNodes.map((item) => item.nodeKey),
    );
    proposal.researchQuestions.forEach((question, index) => {
      if (!perspectiveKeys.has(question.perspectiveKey)) {
        context.addIssue({
          code: "custom",
          path: ["researchQuestions", index, "perspectiveKey"],
          message: "Research question references an unknown perspective candidate.",
        });
      }
    });
    proposal.searchPlans.forEach((search, index) => {
      if (!questionKeys.has(search.questionKey)) {
        context.addIssue({
          code: "custom",
          path: ["searchPlans", index, "questionKey"],
          message: "Search plan references an unknown research question candidate.",
        });
      }
    });
    proposal.knowledgeNodes.forEach((node, index) => {
      if (node.parentNodeKey && !knowledgeKeys.has(node.parentNodeKey)) {
        context.addIssue({
          code: "custom",
          path: ["knowledgeNodes", index, "parentNodeKey"],
          message: "Knowledge node references an unknown parent candidate.",
        });
      }
      node.supportingSourceCandidateKeys.forEach((sourceKey) => {
        if (!sourceKeys.has(sourceKey)) {
          context.addIssue({
            code: "custom",
            path: ["knowledgeNodes", index, "supportingSourceCandidateKeys"],
            message: `Knowledge node references unknown source candidate '${sourceKey}'.`,
          });
        }
      });
    });
    proposal.outlineCandidates.forEach((outline, outlineIndex) => {
      outline.sections.forEach((section, sectionIndex) => {
        section.questionKeys.forEach((questionKey) => {
          if (!questionKeys.has(questionKey)) {
            context.addIssue({
              code: "custom",
              path: ["outlineCandidates", outlineIndex, "sections", sectionIndex, "questionKeys"],
              message: `Outline references unknown question candidate '${questionKey}'.`,
            });
          }
        });
        section.supportingSourceCandidateKeys.forEach((sourceKey) => {
          if (!sourceKeys.has(sourceKey)) {
            context.addIssue({
              code: "custom",
              path: [
                "outlineCandidates",
                outlineIndex,
                "sections",
                sectionIndex,
                "supportingSourceCandidateKeys",
              ],
              message: `Outline references unknown source candidate '${sourceKey}'.`,
            });
          }
        });
      });
    });
    if (
      proposal.status === "complete" &&
      (proposal.perspectives.length === 0 ||
        proposal.researchQuestions.length === 0 ||
        proposal.outlineCandidates.length === 0)
    ) {
      context.addIssue({
        code: "custom",
        path: ["status"],
        message: "A complete exploration requires perspectives, questions, and an outline candidate.",
      });
    }
  });

export const ResearchExplorationStatusSchema = z.enum([
  "queued",
  "running",
  "partial",
  "complete",
  "failed",
  "unknown_outcome",
  "expired",
  "cancelled",
]);

export const ResearchExplorationFailureSchema = z
  .object({
    code: IdentifierSchema,
    category: z.enum(["contract", "provider", "infrastructure", "unknown_outcome"]),
    retryability: z.enum(["none", "safe", "unknown"]),
    technicalMessage: z.string().trim().min(1).max(8_000),
    userMessageCode: IdentifierSchema,
  })
  .strict();

export const ResearchExplorationExecutionSchema = z
  .object({
    schemaVersion: z.literal(1),
    executionId: z.uuid(),
    explorationId: IdentifierSchema,
    explorationRevision: z.number().int().positive(),
    executionRevision: z.number().int().nonnegative(),
    jobId: z.uuid().optional(),
    contextRevision: z.number().int().nonnegative().optional(),
    requirement: z.enum(["optional", "required"]),
    adapter: z.literal("storm"),
    versions: ResearchExplorationVersionSnapshotSchema,
    inputFingerprint: Sha256Schema,
    status: ResearchExplorationStatusSchema,
    remoteExecutionId: IdentifierSchema.optional(),
    resultLocation: z.string().trim().min(1).max(2_000).optional(),
    nextCheckAt: DateTimeSchema.optional(),
    inspectionCount: z.number().int().nonnegative(),
    maximumInspectionCount: z.number().int().positive(),
    expiresAt: DateTimeSchema,
    failure: ResearchExplorationFailureSchema.optional(),
    createdAt: DateTimeSchema,
    updatedAt: DateTimeSchema,
  })
  .strict()
  .superRefine((execution, context) => {
    if (execution.inspectionCount > execution.maximumInspectionCount) {
      context.addIssue({
        code: "custom",
        path: ["inspectionCount"],
        message: "Inspection count exceeds the frozen maximum.",
      });
    }
    if (
      ["running", "partial", "complete"].includes(execution.status) &&
      !execution.remoteExecutionId
    ) {
      context.addIssue({
        code: "custom",
        path: ["remoteExecutionId"],
        message: "Remote execution ID is required after the provider accepts a task.",
      });
    }
    if (["partial", "complete"].includes(execution.status) && !execution.resultLocation) {
      context.addIssue({
        code: "custom",
        path: ["resultLocation"],
        message: "Completed or partial exploration requires a durable result location.",
      });
    }
    if (["failed", "unknown_outcome"].includes(execution.status) && !execution.failure) {
      context.addIssue({
        code: "custom",
        path: ["failure"],
        message: "Failed or unknown exploration requires structured failure evidence.",
      });
    }
  });

export type ResearchExplorationInput = z.infer<typeof ResearchExplorationInputSchema>;
export type ResearchExplorationProposal = z.infer<typeof ResearchExplorationProposalSchema>;
export type ResearchExplorationStatus = z.infer<typeof ResearchExplorationStatusSchema>;
export type ResearchExplorationExecution = z.infer<typeof ResearchExplorationExecutionSchema>;
export type ResearchExplorationVersionSnapshot = z.infer<
  typeof ResearchExplorationVersionSnapshotSchema
>;
